src/smoltcp_phy.rs

```diff
@@ -0,0 +1,106 @@
+//! ENC28J60 wrapper for use as a smoltcp interface
+
+use crate::Enc28j60;
+use embedded_hal::{blocking, digital::v2::OutputPin};
+use smoltcp::{
+    phy::{self, Device, DeviceCapabilities},
+    time::Instant,
+};
+
+/// Wrapper for use as a `smoltcp` interface for sending and receiving raw network frames.
+pub struct Phy<'a, SPI, NCS, INT, RESET> {
+    phy: Enc28j60<SPI, NCS, INT, RESET>,
+    rx_buf: &'a mut [u8],
+    tx_buf: &'a mut [u8],
+}
+
+impl<'a, SPI, NCS, INT, RESET> Phy<'a, SPI, NCS, INT, RESET> {
+    /// Create a new ethernet interface from an Enc28j60, a receive buffer and a transmit buffer.
+    pub fn new(
+        phy: Enc28j60<SPI, NCS, INT, RESET>,
+        rx_buf: &'a mut [u8],
+        tx_buf: &'a mut [u8],
+    ) -> Self {
+        Phy {
+            phy,
+            rx_buf,
+            tx_buf,
+        }
+    }
+}
+
+impl<'a, 'b, E, SPI: 'a, NCS: 'a, INT, RESET> Device<'a> for &mut Phy<'b, SPI, NCS, INT, RESET>
+where
+    SPI: blocking::spi::Transfer<u8, Error = E> + blocking::spi::Write<u8, Error = E>,
+    NCS: OutputPin,
+    INT: crate::sealed::IntPin,
+    RESET: crate::sealed::ResetPin,
+{
+    type RxToken = RxToken<'a>;
+    type TxToken = TxToken<'a, SPI, NCS, INT, RESET>;
+
+    fn receive(&'a mut self) -> Option<(Self::RxToken, Self::TxToken)> {
+        match self.phy.next_packet().ok() {
+            Some(Some(packet)) => {
+                packet.read(&mut self.rx_buf[..]).ok().unwrap();
```

Comment:
I think we have to use the unwrapped value here, otherwise we're discarding the length of the read packet.